handler/src/main/java/io/netty/handler/ssl/SniHandler.java

```diff
@@ -305,19 +305,41 @@ public void operationComplete(Future<SslContext> future) throws Exception {
         }
     }
 
-    private void replaceHandler(ChannelHandlerContext ctx, Selection selection) {
-        SSLEngine sslEngine = null;
-        this.selection = selection;
+    /**
+     * Called upon successful completion of the {@link AsyncMapping}'s {@link Future}.
+     *
+     * @see #select(ChannelHandlerContext, String)
+     */
+    private void onSslContext(ChannelHandlerContext ctx, String hostname, SslContext sslContext) {
+        this.selection = new Selection(sslContext, hostname);
         try {
-            sslEngine = selection.context.newEngine(ctx.alloc());
-            ctx.pipeline().replace(this, SslHandler.class.getName(), selection.context.newHandler(sslEngine));
+            replaceHandler(ctx, hostname, sslContext);
         } catch (Throwable cause) {
             this.selection = EMPTY_SELECTION;
+            ctx.fireExceptionCaught(cause);
+        }
+    }
+
+    /**
+     * The default implementation of this method will simply replace {@code this} {@link SniHandler}
+     * instance with a {@link SslHandler}. Users may override this method to implement custom behavior.
+     *
+     * Please be aware that this method may get called after a client has already disconnected and
+     * custom implementations must take it into consideration when overriding this method.
+     *
+     * It's also possible for the hostname argument to be {@code null}.
+     */
+    protected void replaceHandler(ChannelHandlerContext ctx, String hostname, SslContext sslContext) throws Exception {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "hostname". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)